src/sentry/filters/preprocess_hashes.py

```diff
@@ -11,3 +16,153 @@
+
+
+def get_frame_hash(platform, frame):
+    output = []
+    # Safari throws [native code] frames in for calls like ``forEach``
+    # whereas Chrome ignores these. Let's remove it from the hashing algo
+    # so that they're more likely to group together
+    if frame.get('filename') == '[native code]':
+        return output
+
+    if frame.get('module'):
+        if is_unhashable_module(frame['module']):
+            output.append('<module>')
+        else:
+            output.append(remove_module_outliers(frame['module']))
+    elif frame.get('filename'
+                   ) and not is_url(frame['filename']) and not is_caused_by(frame['filename']):
+        output.append(remove_filename_outliers(frame['filename'], platform))
+
+    context_line = frame.get('context_line')
+    abs_path = frame.get('abs_path', '')
+    if context_line is None:
+        can_use_context = False
+    elif len(context_line) > 120:
+        can_use_context = False
+    elif abs_path.startswith('blob:') or is_url(abs_path) and not frame['function']:
+        # the context is too risky to use here as it could be something
+        # coming from an HTML page or it could be minified/unparseable
+        # code, so lets defer to other lesser heuristics (like lineno)
+        can_use_context = False
+    else:
+        can_use_context = True
+
+    if can_use_context:
+        output.append(context_line)
+    elif not output:
+        # If we were unable to achieve any context at this point
+        # (likely due to a bad JavaScript error) we should just
+        # bail on recording this frame
+        return output
+    elif frame.get('symbol'):
+        output.append(frame.get('symbol'))
+    elif frame.get('function'):
+        if is_unhashable_function(frame['function']):
+            output.append('<function>')
+        else:
+            output.append(remove_function_outliers(frame['function']))
+    elif frame.get('lineno') is not None and frame.get('colno') is not None:
+        output.extend([frame['lineno'], frame['colno']])
+    return output
+
+
+def get_stacktrace_hash(platform, stacktrace):
+    frames = stacktrace['frames']
+
+    stack_invalid = is_invalid_stack(frames)
+    if stack_invalid:
+        return []
+
+    output = []
+    for frame in frames:
+        output.extend(get_frame_hash(platform, frame))
+
+    return output
+
+
+def get_exception_hash(platform, exception):
+    output = []
+    for value in exception['values']:
+        if not value['stacktrace']:
+            continue
+        stack_hash = get_stacktrace_hash(platform, value['stacktrace'])
+        if stack_hash:
+            output.extend(stack_hash)
+            output.append(value.get('type'))
+
+    if not output:
+        for value in exception['values']:
+            output.extend([v for v in [value.get('type'), value.get('value')] if v])
+
+    return output
+
+
+def get_preprocess_defaults(data):
```

Comment:
It probably makes sense to replicate the normal fingerprint logic here and only use the "highest scoring" interface, rather than collapsing them all into the hash inputs.

Reply:
Also, this should probably be `get_preprocess_hash_inputs` or something like we had talked about.